Packet++/src/RawPacket.cpp

```
@@ -136,6 +136,13 @@ namespace pcpp
 
 	void RawPacket::insertData(int atIndex, const uint8_t* dataToInsert, size_t dataToInsertLen)
 	{
+		// Check for overflow in the new length
+		if ((size_t)m_RawDataLen + dataToInsertLen < (size_t)m_RawDataLen)
+		{
+			PCPP_LOG_ERROR("RawPacket::insertData: dataToInsertLen causes overflow");
+			return;
+		}
```

Can we throw an `std::length_error` instead of just returning with a error message? The error message will not indicate the error condition to the caller.

---

Will throwing an exception prevent the application from crashing? 🤔 

This method was called in `BgpLayer` so we should probably think how to prevent it from being called with malformed data

---

You want a crash if overflow happens. Or at least you want an exception thrown to indicate the failure of the operation, and if the caller doesn't handle it, then to crash. It isn't really possible to recover from the situation as the system has reached the limit of how much data can be stored.

Silently continuing with just an error message in the log is worse as the program will assume the operation went file and probably produce corrupted data down the line. This is exactly the type of improbable situation the exception mechanism was made to handle.

---

I totally agree, however I don't think throwing an exception will resolve the OSS-Fuzz issue. We should probably throw an exception and fix the issue that caused it in `BgpLayer`

---

Hi, just to confirm - I will update the patch to throw an `std::length_error` . Would that be accepted by both of you? @Dimi1010 @seladb

---

@Marti2203 we can throw an `std::length_error` but should also fix the issue in `BgpLayer`. This exception alone won't solve the issue because the program will still crash because of the exception

---

Sure, will prepare a patch for that

---

Hi @seladb @Dimi1010 apologies for the delay, I have pushed a commit that I think handles it as discussed - an error is thrown from the raw packet which is caught in the `BGPLayer` to convert into a bool.

---

@Marti2203 thanks for working on it!
Catching exceptions is pretty expensive in C++. I'd avoid using `try..catch` and instead check the sizes and prevent calling `extendLayer` is we know it'd fail

---

@seladb @Marti2203 

Generally, I find exceptions are fine as long as they aren't used to dictate common control flow. The current mainstream compilers use an architecture that has a "zero cost" try-catch blocks if the exception isn't thrown. The tradeoff is a significant overhead if an exception is thrown, so you only want them thrown for **rare** events.

The current usage seems ok to me? How often are overflows expected to happen?

---

Hi, I have been working on other things and have been checking out this PR again. I agree with sela's points and will prepare a patch after a confirmation from your side - 
* NO exception will be thrown in RawPacket, only it is logged and an error is returned. This is to ensure new clients of the code do not lead to a crash and code stylistically aligns with the rest. 
* Separately, will add a check for the error in the BGPLayer as it is the only client in the project that uses it.

---

I think we can still throw an error in `insertData()` but I would avoid `try..catch` in `BgpLayer`  and instead check the sizes before calling `extendLayer()`. @Marti2203 let me know what you think

---

Sounds good. I will also cover the nitpick and send it tonight :)